neuralmonkey/dataset.py

```diff
@@ -0,0 +1,559 @@
+        assert self.length is not None
+        return self.length
+
+    @property
+    def series(self) -> List[str]:
+        return list(sorted(self.iterators.keys()))
+
+    def has_series(self, name: str) -> bool:
+        """Check if the dataset contains a series of a given name.
+
+        Arguments:
+            name: Series name
+
+        Returns:
+            True if the dataset contains the series, False otherwise.
+        """
+        return name in self.iterators
+
+    def get_series(self, name: str) -> Iterator:
+        """Get the data series with a given name.
+
+        Arguments:
+            name: The name of the series to fetch.
+
+        Returns:
+            A freshly initialized iterator over the data series.
+
+        Raises:
+            KeyError if the series does not exists.
+        """
+        return self.iterators[name]()
+
+    def maybe_get_series(self, name: str) -> Optional[Iterator]:
+        """Get the data series with a given name, if it exists.
+
+        Arguments:
+            name: The name of the series to fetch.
+
+        Returns:
+            The data series or None if it does not exist.
+        """
+        if name in self.iterators:
+            return self.get_series(name)
+        return None
+
+    def batches(self, batch_size: int) -> Iterator["Dataset"]:
+        """Split the dataset into batches.
+
+        Arguments:
+            batch_size: The size of a batch.
+
+        Returns:
+            Generator yielding the batches.
+        """
+        if self.buffer_size is not None and self.buffer_size < batch_size:
+            warn("Buffer size ({}) is lower than batch size ({})."
+                 .format(self.buffer_size, batch_size))
+
+        # Initialize iterators
+        iterators = {s: it() for s, it in self.iterators.items()}
+
+        # Create iterator over instances
+        zipped_iterator = (
+            dict(zip(iterators, row)) for row in zip(*iterators.values()))
+
+        # Fill the buffer with initial values, shuffle optionally
+        if self.buffer_size is not None:
+            # pylint: disable=stop-iteration-return
+            # This is pylint issue https://github.com/PyCQA/pylint/issues/2158
+            buf = deque(next(zipped_iterator) for _ in range(self.buffer_size))
+            # pylint: enable=stop-iteration-return
+        else:
+            assert not self.lazy
+            buf = deque(zipped_iterator)
+        if self.shuffled:
+            random.shuffle(buf)  # type: ignore
+
+        def _make_datagen(rows, key):
+            def itergen():
+                return (row[key] for row in rows)
+            return itergen
+
+        # Iterate over the rest of the data until buffer is empty
+        batch_index = 0
+        while buf:
+            # Create the batch
+            name = "{}.batch.{}".format(self.name, batch_index)
+            rows = [buf.popleft() for _ in range(batch_size) if buf]
```

Comment:
It should probably be mentioned in the documentation that it returns batches of size == buffer_size if buffer_size < batch_size

Reply:
to je bug, kterej nastane v extrémním případě. je tam warning, ale radši to nějak opravim, aby to vždycky dávalo batch_size

Reply:
Ono je vůbec otázka, jestli to chceme plnit a shufflovat po každý batchi, nebo jestli se tam nemá dát nějakej lower threshold pro velikost bufferu, pod kterej když to vleze, tak se to doplní na buffer_size

Reply:
Taky shufflování na deque je kvadratický, protože neni random access

Reply:
Navíc nevidim, proč jsem zvolil zrovna `deque`, když na tom volám jen `popleft`, `shuffle`, `extend`, `len` a testuju na prázdnotu

Reply:
ale jo, když se neshuffluje tak to tam chceš rvát jednou stranou a druhou to házet ven

Reply:
So, I propose:

1. Instead of a  warning, raise an exception, that the buffer is too small.
2. Document that in the docstring
3. Introduce another option `buffer_min_size`, that would be by default half the size of the buffer. (Think buffer_size of 1000 sentences, after 10 batches of 50 sentences each which we take from the buffer, the buffer gets refilled and reshuffled.)

Reply:
Anebo 1 škrtnout a přidat jen dokumentaci.

Reply:
htoovo